magento/magento2
# MC-41852: issue with product image cache

### Description (*)
This PR fixes issue when product image that contains dot in image name at first or second position (e.g. "i.mage.png") is not showing on frontend after flushing catalog image cache.

This happened because when uploading product image magento uses files dispersion setting and saving image to path generated by `\Magento\Framework\File\Uploader::getDispersionPath` which consists of two directories defined by two first letters of filename (e.g. path for "image.png" will be /i/m/image.png). But, in case when there is dot in first two letters of filename, directory name with underscore will be used (e.g. path for filename "i.mage.png" will be /i/_/image.png).

When catalog image cache is flushed and frontend product image is requested magento tries to create new cache file in `\Magento\MediaStorage\App\Media::createLocalCopy` and tries to retrieve original image location by calling `\Magento\MediaStorage\App\Media::getOriginalImage`.

This function is using regexp to get from path like _"media/catalog/product/cache/838ef3af6b51df9f635b7eb110d280c6/i/m/image.png"_ original image path _"/i/m/image.png"_. 
This regexp was changed in https://github.com/magento/magento2/commit/362eba826be3a7aff9ce9de579893e639ffc2cfd from `|^.*((?:/[^/]+){3})$|` to `|^.*?((?:/([^/])/([^/])/\2\3)?/?[^/]+$)|`. New regexp (`|^.*?((?:/([^/])/([^/])/\2\3)?/?[^/]+$)|`) is not taking into consideration cases when original filename contained dot in first two letters and therefore path with underscore instead of dot is generated.

In this PR old regexp (`|^.*((?:/[^/]+){3})$|`) was restored since its working correctly with all filenames.

### Related Pull Requests


### Fixed Issues (if relevant)

1. Fixes magento/magento2#32699

### Manual testing scenarios (*)

1. Create a new product or go to the existing one
2. Add image to product (image name should contain dot in first or second position, e.g. "i.mage.png")
3. Save product
4. Go to System -> Cache Management and click button "Flush Catalog Image Cache"
5. Open product on frontend

**Actual result:** (before this PR)
Placeholder showed instead of product image

**Expected result:** (after this PR)
Product image is showed

### Questions or comments

Since commit https://github.com/magento/magento2/commit/362eba826be3a7aff9ce9de579893e639ffc2cfd that changed working regexp to not correctly working with all filenames regexp was introduced with "[performance]" text in commit message, it was probably fixing some performance issues of older regexp.
In this PR it was decided not to modify this not fully working regexp but rather restore old one because it was not very clear what performance changes is was making and because of possible risks of introducing new regexp.
So, reviewer probably should take a close look at deleted regexp before accepting this PR.


### Contribution checklist (*)
